lib/harmony/harmonytask/taskTypeHandler.go

```diff
@@ -0,0 +1,274 @@
+package harmonytask
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var logger = logging.Logger("harmonytask")
+
+type taskTypeHandler struct {
+	TaskInterface
+	TaskTypeDetails
+	TaskEngine *TaskEngine
+	Count      int /// locked by TaskEngine's mutex
+
+	LastCleanup atomic.Value
+}
+
+func (h *taskTypeHandler) AddTask(extra func(TaskID, *harmonydb.Tx) bool) {
+	var tID TaskID
+	did, err := h.TaskEngine.db.BeginTransaction(h.TaskEngine.ctx, func(tx *harmonydb.Tx) bool {
+		// create taskID (from DB)
+		_, err := tx.Exec(`INSERT INTO harmony_task (name, added_by, posted_time) 
+			VALUES ($1, $2, CURRENT_TIMESTAMP) `, h.Name, h.TaskEngine.ownerID)
+		if err != nil {
+			logger.Error("Could not insert into harmonyTask", err)
+			return false
+		}
+		err = tx.QueryRow("SELECT id FROM harmony_task ORDER BY update_time DESC LIMIT 1").Scan(&tID)
+		if err != nil {
+			logger.Error("Could not select ID: ", err)
+		}
+		return extra(tID, tx)
+	})
+	if err != nil {
+		logger.Error(err)
+	}
+	if !did {
+		return
+	}
+
+	if !h.considerWork([]TaskID{tID}) {
+		h.TaskEngine.bump(h.Name) // We can't do it. How about someone else.
+	}
+}
+
+func (h *taskTypeHandler) considerWork(ids []TaskID) (workAccepted bool) {
+	if len(ids) == 0 {
+		return true // stop looking for takers
+	}
+
+	h.TaskEngine.workAdderMutex.Lock()
+	defer h.TaskEngine.workAdderMutex.Unlock()
+
+	// 1. Can we do any more of this task type?
+	if h.Max > -1 && h.Count == h.Max {
+		logger.Info("Did not accept " + h.Name + " task: at max already.")
```

Comment:
Would be great to have these use the structured-logging calls, e.g.
```suggestion
		logger.Infow("did not accept task", "type", h.Name, "reason", "at max already.")
```

Most deployments will dump their logs at things like logstash, and structured logs are vastly easier to query and make pretty Grafana dashboards from